hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -0,0 +1,402 @@
+package nodepool
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	api "github.com/openshift/hypershift/api"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	hyperutil "github.com/openshift/hypershift/hypershift-operator/controllers/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	k8sutilspointer "k8s.io/utils/pointer"
+	capiv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// reconcileInPlaceUpgrade loops over all Nodes that belong to a NodePool and performs an in place upgrade if necessary.
+func (r *NodePoolReconciler) reconcileInPlaceUpgrade(ctx context.Context, hc *hyperv1.HostedCluster, nodePool *hyperv1.NodePool, machineSet *capiv1.MachineSet, targetConfigHash, targetVersion, targetConfigVersionHash string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// If there's no guest cluster yet return early.
+	if hc.Status.KubeConfig == nil {
+		return nil
+	}
+
+	hostedClusterClient, err := newHostedClusterClient(ctx, r.Client, hc)
+	if err != nil {
+		return fmt.Errorf("failed to create remote client: %v", err)
+	}
+
+	// Watch hosted cluster Nodes. We track the created caches, so we don't add a watcher on every reconciliation.
+	r.hostedClusterCachesTracker.Lock()
+	defer r.hostedClusterCachesTracker.Unlock()
+	if !r.hostedClusterCachesTracker.caches[client.ObjectKeyFromObject(nodePool)] {
+		hostedClusterCache, err := newHostedClusterCache(ctx, r.Client, hc)
+		if err != nil {
+			return fmt.Errorf("failed to create hosted cluster cache: %v", err)
+		}
+
+		// TODO (alberto): cancel the ctx on exit.
```

Comment:
ack, makes sense if you create a new context for the cache and store the cancel function in the map